src/FffGcodeWriter.cpp

```diff
@@ -1660,57 +1732,86 @@ bool FffGcodeWriter::addSupportInfillToGCode(const SliceDataStorage& storage, La
                 gcode_layer.addPolygonsByOptimizer(boundary, &gcode_layer.configs_storage.support_infill_config);
             }
-            offset_from_outline = -support_line_width;
-            support_infill_overlap = infill_extr_here.getSettingInMicrons("infill_overlap_mm"); // support lines area should be expanded outward to overlap with the boundary polygon
         }
 
-        int extra_infill_shift = 0;
-        bool use_endpieces = true;
-        Polygons* perimeter_gaps = nullptr;
-        double fill_angle = 0;
-        Infill infill_comp(support_pattern, island, offset_from_outline, support_line_width, support_line_distance, support_infill_overlap, fill_angle, z, extra_infill_shift, perimeter_gaps, infill_extr.getSettingBoolean("support_connect_zigzags"), use_endpieces);
-        Polygons support_polygons;
-        Polygons support_lines;
-        infill_comp.generate(support_polygons, support_lines);
-        if (support_lines.size() > 0 || support_polygons.size() > 0)
+        // process sub-areas in this support infill area with different densities
+        for (unsigned int density_idx = 0; density_idx < support_infill_part.gradual_infill_areas_per_combine_per_density.size(); ++density_idx)
         {
-            setExtruder_addPrime(storage, gcode_layer, layer_nr, infill_extruder_nr_here); // only switch extruder if we're sure we're going to switch
-            gcode_layer.setIsInside(false); // going to print stuff outside print object, i.e. support
-            gcode_layer.addPolygonsByOptimizer(support_polygons, &gcode_layer.configs_storage.support_infill_config);
-            gcode_layer.addLinesByOptimizer(support_lines, &gcode_layer.configs_storage.support_infill_config, (support_pattern == EFillMethod::ZIG_ZAG)? SpaceFillType::PolyLines : SpaceFillType::Lines);
-            added = true;
+            if (support_infill_part.gradual_infill_areas_per_combine_per_density[density_idx].empty())
+            {
+                continue;
+            }
+            const Polygons& support_area = support_infill_part.gradual_infill_areas_per_combine_per_density[density_idx][0];
+
+            unsigned int density_factor = 2 << density_idx; // == pow(2, density_idx + 1)
+            int support_line_distance_here = support_line_distance * density_factor; // the highest density infill combines with the next to create a grid with density_factor 1
+            int support_shift = support_line_distance_here / 2;
+            if (density_idx == support_infill_part.gradual_infill_areas_per_combine_per_density.size() - 1)
+            {
+                support_line_distance_here /= 2;
+            }
+
+            int offset_from_outline = 0;
+            int support_infill_overlap = 0; // support infill should not be expanded outward
+
+            if (support_pattern == EFillMethod::GRID || support_pattern == EFillMethod::TRIANGLES || support_pattern == EFillMethod::CONCENTRIC)
```

Comment:
Fixed in ad0677c801591da4fd6064cb7114f0d51fb8bcdc
There is no need to check `part.insets.size() > 0` because it won't reach this part of code in the loop in the first place.